lib/rtp_vp9/vp9_payload_descriptor.ex

```diff
@@ -0,0 +1,347 @@
+  ```
+  """
+
+  @type first_octet :: 0..255
+
+  @type picture_id :: 0..32767
+
+  @type tid :: 0..7
+
+  @type u :: 0..1
+
+  @type d :: 0..1
+
+  @type sid :: 0..7
+
+  @type p_diff :: 0..255
+
+  @type tl0picidx :: 0..255
+
+  defmodule PGDescription do
+    alias Membrane.RTP.VP9.PayloadDescriptor
+
+    @type t :: %__MODULE__{
+            tid: PayloadDescriptor.tid(),
+            u: PayloadDescriptor.u(),
+            p_diffs: list(PayloadDescriptor.p_diff())
+          }
+
+    defstruct [:tid, :u, p_diffs: []]
+  end
+
+  defmodule SSDimension do
+    @type t :: %__MODULE__{
+            width: 0..65535,
+            height: 0..65535
+          }
+
+    @enforce_keys [:width, :height]
+    defstruct @enforce_keys
+  end
+
+  defmodule ScalabilityStructure do
+    alias Membrane.RTP.VP9.PayloadDescriptor
+    alias Membrane.RTP.VP9.PayloadDescriptor.{SSDimension, PGDescription}
+
+    @type t :: %__MODULE__{
+            first_octet: PayloadDescriptor.first_octet(),
+            dimensions: list(SSDimension.t()),
+            pg_descriptions: list(PGDescription.t())
+          }
+
+    @enforce_keys [:first_octet]
+    defstruct [:first_octet, dimensions: [], pg_descriptions: []]
+  end
+
+  @type t :: %__MODULE__{
+          first_octet: first_octet(),
+          picture_id: picture_id(),
+          tid: tid(),
+          u: u(),
+          sid: sid(),
+          d: d(),
+          p_diffs: list(p_diff()),
+          tl0picidx: tl0picidx(),
+          scalability_structure: ScalabilityStructure.t()
+        }
+
+  defstruct [
+    :first_octet,
+    :picture_id,
+    :tid,
+    :u,
+    :sid,
+    :d,
+    :tl0picidx,
+    :scalability_structure,
+    p_diffs: []
+  ]
+
+  @spec parse_payload_descriptor(binary()) :: {:error, :malformed_data} | {:ok, {t(), binary()}}
+  def parse_payload_descriptor(raw_payload)
+
+  def parse_payload_descriptor(<<header::binary-size(1), rest::binary()>>)
+      when byte_size(rest) > 0 do
+    <<i::1, _p::1, _l::1, f::1, _bevz::4>> = header
+
+    if i != f do
+      {:error, :malformed_data}
+    else
+      case get_pid(header, rest, %__MODULE__{first_octet: :binary.decode_unsigned(header)}) do
+        {:ok, {descriptor_acc, rest}} ->
+          case get_layer_indices(header, rest, descriptor_acc) do
+            {:ok, {descriptor_acc, rest}} ->
+              case get_pdiffs(header, rest, 0, descriptor_acc) do
+                {:ok, {descriptor_acc, rest}} ->
+                  case get_scalability_structure(header, rest) do
+                    {:ok, {ss, rest}} ->
+                      {:ok, {%{descriptor_acc | scalability_structure: ss}, rest}}
+
+                    _ ->
+                      {:error, :malformed_data}
+                  end
+
+                _ ->
+                  {:error, :malformed_data}
+              end
+
+            _ ->
+              {:error, :malformed_data}
+          end
+
+        _ ->
+          {:error, :malformed_data}
+      end
+    end
```

Comment:
I know Christmas is coming, but please don't place Christmas trees in the code ;) Use `with` instead